### ISSUE TITLE
Guarantee an upper limit for required free space

### ISSUE THREAD

**Issue by fulldecent** [borgbackup/borg#1225]
Currently, the advice is:

> Before you start creating backups, please make sure that there is always a good amount of free space on the filesystem that has your backup repository (and also on ~/.cache). It is hard to tell how much, maybe 1-5%.
>
> If you run out of disk space, it can be hard or impossible to free space, because Borg needs free space to operate - even to delete backup archives. There is a --save-space option for some commands, but even with that Borg will need free space to operate.

There is no specific prescription ("it is hard to tell how much"), and noncompliance is disastrous (the backup is unusable). This is a bad combination.

Please create a specific, calculable UPPER BOUND for the prescribed required free space. This can be much higher than the actual required free space, but there should be no circumstance imaginable where more than the UPPER BOUND is required.

*Related: #1224 discusses architecture documentation. I would be able to answer this question if I understood the architecture.*

 - [ ] TO BE A NEW ISSUE: Update documentation at https://borgbackup.readthedocs.io/en/stable/quickstart.html#important-note-about-free-space to provide specific free space upper bound
 - [ ] TO BE A NEW ISSUE: precalculate free space requirements before backing up and abort quickly if the threshold will be passed
 - [ ] TO BE A NEW ISSUE: auto prune old backups when backing up (Time Machine has this feature)

**Comment by ThomasWaldmann** [comment]
Docs: Maybe a formula for an upper bound could be determined and documented, it is just some work, depends on many factors and not very exciting, so consider whether a bounty would be good for this (or if you want to work on that).

Aborting: This is a single-pass backup program. At the start of a backup, it does not even know how much space it will need for this backup (nor how many files there are, nor how big the files are, nor how good they dedup or compress, nor how much metadata there is). So, it can not abort before starting to backup files. It could abort somewhere in the middle (maybe creating a checkpoint), when available space gets below the threshold (it would need to poll free space now and then for that).

Auto-Prune: If you want that, you just put "borg prune" additionally into your script file that does the backup.

In general: please keep issues clean, this one looks a bit like 3 issues mixed into 1.

**Comment by fulldecent** [comment]
Sorry, to clarify, the three items at the bottom are potential follow up if there is a good upper bound.

**Comment by enkore** [comment]
Additional space usage when deleting w/ save space: Borg needs to be able to journal all object deletions plus at most one segment (5..10 MB / ~500 MB) during compaction, plus ~one index size when committing (see docs for formula of that; 40 * object count). Each deletion is 41 bytes plus some sub-byte/deletion overhead. Say you want to delete an archive with 123456 unique(!) chunks, then you'll need some 5 MB to journal the deletes, some 10 MB during compaction and one index size. Say you got 15 million chunks total, the index is ~600 MB.

Re. free space handling: this is an area that's notoriously difficult to handle, because you can't even create a file if your quota is full. That makes basic operations hard / impossible in many cases. This is also the reason why traditionally some percentage of blocks are reserved for root on \*nix. However, if you're dumping tons of data to disks *as root* then you have to handle that block reservation yourself.

**Comment by fulldecent** [comment]
Even if we can say something like you need to keep this much free:

 > 505 MB + 41 bytes * number of files

This would be useful because we can document that. But also we can check for that BEFORE doing an operation and present a warning. And that would allow us make a much better statement in the BIG RED SCARY WARNING.

**Comment by ThomasWaldmann** [comment]
For a simple and save **--save-space required free space upper bound formula**:

The repo index is a hashtable, so its size is not just 40B * chunk count, but also one has to add bigger amounts of free space in it. But if we take on-disk repo index size as a known value, we could do an estimation based on it:

- we need ~ 1 index size free for the transaction
- as the deletes are 41 bytes and the repo index entries are 40 bytes, we could say we also need about another ~ 1 index size for all the deletes (if we deleted all chunks)
- if there could be a circumstance that triggers hash table growth, the new one could grow up to 2x
- 10MB (1.0) or 500MB (1.1) for 1 segment

So, how about "4x index size" but not less than 100MB (1.0) / 2GB (1.1)?

**Comment by varac** [comment]
I filled my backup partition with borg and now even borg list is crashing. So i'm very interested in the `precalculate free space requirements before backing up and abort quickly if the threshold will be passed` feature !

**Comment by enkore** [comment]
Try this script

```py
import sys

from borg.repository import Repository

with Repository(sys.argv[1], lock=False) as repository:
    lio = repository.io
    for segment, segment_file in lio.segment_iterator():
        if lio.is_committed_segment(segment):
            print('Committed', segment_file)
```

It doesn't lock so doesn't write and should work on a full disk. It should give you a list of "committed" segments. If your repository has segments beyond the last committed segment you can rollback manually by deleting these.

(If you have the master branch installed, make it `is_committed_segment(segment_file)` instead)

**Comment by enkore** [comment]
I'm closing this in favour of the more specific tickets (if you feel that one is missing, please create one): #1272 #1273 #1262 

As discussed above the headline topic won't happen, because we won't make borg (create) a two-pass backup program for many good reasons, and therefore can never know beforehand how much space will be needed to fit a backup program. With the recent fix in 1.0.4 (#1263) this isn't a real issue anymore anyway for almost any setting -- except very low transfer rates or very low checkpoint intervals (you people know who you are ;)
